### PR TITLE
Fix memory leaks post game exit

### PR DESCRIPTION
**Pull request by peppy**
This is mostly to reduce nUnit memory/thread usage overhead.

In normal game execution everything is freed when the application exits, so this should not present any discernible improvement to an end user.

### CODE REVIEW

File: osu.Framework/Platform/GameHost.cs [hunk 1]
```diff
@@ -614,6 +630,8 @@ protected virtual void Dispose(bool disposing)
 
         public void Dispose()
         {
+            Root = null;
```
Comment (smoogipoo) [inline osu.Framework/Platform/GameHost.cs]: Why is this here and not in `Dispose(disposing)`?

File: osu.Framework/Platform/Windows/WindowsGameHost.cs [hunk 2]
```diff
@@ -25,17 +25,21 @@ internal WindowsGameHost(string gameName, bool bindIPC = false)
...
+        {
+            if (Window.WindowState != OpenTK.WindowState.Minimized)
+                OnActivated();
+            else
+                OnDeactivated();
         }
 
         protected override void Dispose(bool isDisposing)
         {
+            Window.WindowStateChanged -= onWindowOnWindowStateChanged;
```
Comment (smoogipoo) [inline osu.Framework/Platform/Windows/WindowsGameHost.cs]: All the changes in this file shouldn't be necessary anymore right? Since we're disposing the window and the toolkit and shouldn't have a reference held.

File: osu.Framework/Platform/GameHost.cs [hunk 3]
```diff
@@ -599,11 +607,20 @@ protected virtual void Dispose(bool disposing)
             while (executionState > ExecutionState.Stopped)
                 Thread.Sleep(10);
 
+            AppDomain.CurrentDomain.UnhandledException -= exceptionHandler;
+
             Root?.Dispose();
 
+            Activated = null;
+            Deactivated = null;
```
Comment (smoogipoo) [inline osu.Framework/Platform/GameHost.cs]: These likewise shouldn't be necessary anymore right?

File: osu.Framework/Platform/Windows/WindowsGameHost.cs (hunk 2 above)
Comment (peppy) [reply osu.Framework/Platform/Windows/WindowsGameHost.cs]: it's still best practice to do this. there could be unexpected behaviour if, for instance, the window state changes post disposal of the `WindowsGameHost` (it could fire something bound here that is no longer expected to be in context).

File: osu.Framework/Platform/GameHost.cs [hunk 4]
```diff
@@ -33,7 +33,9 @@ namespace osu.Framework.Platform
 {
     public abstract class GameHost : IIpcHost, IDisposable
     {
-        public GameWindow Window;
+        public GameWindow Window { get; protected set; }
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Framework/Platform/GameHost.cs]: no, check usage

[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Framework/Platform/GameHost.cs]: Interesting... Alright let's go with this for now then.

File: osu.Framework/Threading/Scheduler.cs [hunk 5]
```diff
@@ -14,7 +14,7 @@ namespace osu.Framework.Threading
     /// <summary>
     /// Marshals delegates to run from the Scheduler's base thread in a threadsafe manner
     /// </summary>
-    public class Scheduler : IDisposable
```
Comment (Tom94) [inline osu.Framework/Threading/Scheduler.cs]: Not sure I like this change. Consider the case, where a `ThreadedScheduler` is stores as a `Scheduler` member due to polymorphism. Now, there is no way to explicitly dispose the `ThreadedScheduler` despite that being perhaps desired.

Comment (peppy) [reply osu.Framework/Threading/Scheduler.cs]: threaded scheduler is used in only one place. if you disagree here i'd rather remove the threaded behaviour from it completely.

Comment (Tom94) [reply osu.Framework/Threading/Scheduler.cs]: What's so undesirable about keeping `Scheduler` as `IDisposable`?

Comment (peppy) [reply osu.Framework/Threading/Scheduler.cs]: feels weird having something disposable when all but one very private use-case do not need disposal. it forces us to call `.Dispose()` in usages where it's completely unnecessary.

Comment (Tom94) [reply osu.Framework/Threading/Scheduler.cs]: I don't think it does. Disposal is always done by the finalizer anyway, so calling it explicitly is entirely optional and related to deterministic resource de-allocation.

Comment (peppy) [reply osu.Framework/Threading/Scheduler.cs]: and as such, there's already a finaliser on `ThreadedScheduler`.

i'd argue if you're constructing a `ThreadedScheduler`, as the consumer you should be aware of its need for disposal.

Comment (Tom94) [reply osu.Framework/Threading/Scheduler.cs]: As a consumer you often expect the generic `Scheduler` class, not some specialized more specific class (and the implementation details coming with it). Allowing explicit disposal strictly adds optional functionality, so I see no reason against having it.

This is like having an abstract `AudioTrack` class where some derivatives (e.g. `BassAudioTrack`) require disposal. It'd be crazy to not expose the dispose all within `AudioTrack`, because every consumer will be implemented against the generic `AudioTrack` rather than the specific `BassAudioTrack`.

Comment (peppy) [reply osu.Framework/Threading/Scheduler.cs]: In this case, I'm saying I'd rather remove `ThreadedScheduler` (there are better options for threading tasks already) and just make `Scheduler` have a more limited scope.

`ThreadedScheduler` is only used by `Logger` at the moment (and internally uses `Thread`s directly instead of `Task`s, which we're moving away from).

Comment (Tom94) [reply osu.Framework/Threading/Scheduler.cs]: Sounds like a fair resolution to me.

File: osu.Framework/Logging/Logger.cs [hunk 6]
```diff
@@ -400,8 +401,33 @@ public static void Flush()
...
+                }
+
+                int performUpdate() => Storage != null ? scheduler.Update() : 0;
+
+                cancellationToken = new CancellationTokenSource();
+                writerTask = Task.Factory.StartNew(() =>
+                {
+                    try
+                    {
+                        while (!cancellationToken.IsCancellationRequested)
```
Comment (smoogipoo) [inline osu.Framework/Logging/Logger.cs]: This is going to consume one full thread since there no longer is a `Thread.Sleep()` in here (as ThreadedScheduler had).

File: osu.Framework/Logging/Logger.cs [hunk 7]
```diff
@@ -400,8 +401,33 @@ public static void Flush()
         {
             lock (flush_sync_lock)
             {
-                backgroundScheduler.Dispose();
-                backgroundScheduler = new ThreadedScheduler(@"Logger") { Enabled = storage != null };
+                if (writerTask != null)
+                {
+                    cancellationToken.Cancel();
+                    writerTask.Wait(500);
```
Comment (smoogipoo) [inline osu.Framework/Logging/Logger.cs]: Is there a reason why this is waiting for 500ms, vs just `Wait()`? I'd expect `Flush()` to _completely_ flush the log, not potentially only a part of the log.

File: osu.Framework/Logging/Logger.cs [hunk 8]
```diff
@@ -400,8 +401,33 @@ public static void Flush()
...
+                writerTask = Task.Factory.StartNew(() =>
+                {
+                    try
+                    {
+                        while (!cancellationToken.IsCancellationRequested)
+                            performUpdate();
+                    }
+                    catch (TaskCanceledException)
+                    {
+                    }
```
Comment (smoogipoo) [inline osu.Framework/Logging/Logger.cs]: This catch should never occur, since nothing in the try { } will throw this exception.

Comment (smoogipoo) [reply osu.Framework/Logging/Logger.cs]: This catch should be around the `Wait()`, since that is the method that will give you a task cancelled exception, in the form of an `AggregateException`. This catch being here is potentially problematic since it is passed to the task constructor.

File: osu.Framework/Logging/Logger.cs (hunk 7 above)
Comment (smoogipoo) [reply osu.Framework/Logging/Logger.cs]: We should also be disposing the cancellationToken here (it does not implement a finaliser).

File: osu.Framework/Logging/Logger.cs (hunk 6 above)
Comment (peppy) [reply osu.Framework/Logging/Logger.cs]: good catch

File: osu.Framework/Logging/Logger.cs [hunk 9]
```diff
@@ -390,18 +383,54 @@ private void ensureHeader()
...
+        static Logger()
+        {
+            Flush();
+        }
 
         /// <summary>
         /// Pause execution until all logger writes have completed and file handles have been closed.
+        /// This will also unbind all handlers bound to <see cref="NewEntry"/>.
         /// </summary>
         public static void Flush()
```
Comment (smoogipoo) [inline osu.Framework/Logging/Logger.cs]: Is there a reason why this method is coded the way it is? Creating a new task and all that? Can we have another method like `Close()` or `Finalise()` or `Stop()` as part of Logger that stops the task, and have one task running for the entire thing?

File: osu.Framework/Logging/Logger.cs (hunk 7 above)
Comment (peppy) [reply osu.Framework/Logging/Logger.cs]: 500 was a sane value which allows for writing at least a few megabytes of data. i prefer not having infinite timeouts where possible.

File: osu.Framework/Logging/Logger.cs (hunk 9 above)
Comment (peppy) [reply osu.Framework/Logging/Logger.cs]: Are you saying there should be separate start and stop methods?

Comment (smoogipoo) [reply osu.Framework/Logging/Logger.cs]: The following is the idea I'm going for:

```
static Logger()
{
    task = new Task(() =>
    {
        while (!cancellationToken.IsCancellationRequested)
        {
            if (flushToken.IsCancellationRequested)
                while (performUpdate()) { }
            else
                performUpdate();
            Thread.Sleep(50);
        }
    });
};

public Logger.Flush()
{
    flushToken.Cancel();
    while (scheduler.HasItems) { }
}

internal Logger.Stop()
{
    stopToken.Cancel();
    task.Wait();
}

GameHost.Dispose() => Logger.Flush(); Logger.Stop();
```

Maybe not 100% this since the flush stuff has its own implementation issues, and perhaps it might be better to use a second task in there, but I think the code would be much cleaner this way, and we wouldn't have random tasks/cancellation tokens being constructed on dispose.

File: osu.Framework/Logging/Logger.cs (hunk 7 above)
Comment (peppy) [reply osu.Framework/Logging/Logger.cs]: An example case where this may be a good idea (which is why i added this in the first place) would be if there's a second thread logging new entries faster than the writing can complete and it becomes an endless loop.

File: osu.Framework/Logging/Logger.cs (hunk 9 above)
Comment (peppy) [reply osu.Framework/Logging/Logger.cs]: `Flush` would still have to restart the task. Flush is supposed to keep the logger in an operational state should further logging need to happen (it just ensures we are flushed to a certain point).

Comment (smoogipoo) [reply osu.Framework/Logging/Logger.cs]: Flush wouldn't stop the task in the first place.

File: osu.Framework/Logging/Logger.cs [hunk 10]
```diff
@@ -390,17 +385,36 @@ private void ensureHeader()
...
+        private static readonly ManualResetEvent writer_idle = new ManualResetEvent(true);
+
+        static Logger()
+        {
+            Task.Factory.StartNew(() =>
+            {
+                while (true)
+                {
+                    if ((Storage != null ? scheduler.Update() : 0) == 0)
+                        writer_idle.Set();
```
Comment (smoogipoo) [inline osu.Framework/Logging/Logger.cs]: We need to reset `writer_idle` here:

```
else
    writer_idle.Reset();
```

**Comment by smoogipoo** [comment]
👍 Lets do this 🤞
